### DISCUSSION TITLE
$or and $in operators do not seem to be using indices

### DISCUSSION THREAD

**Discussion by roedoejet** [apache/couchdb#2856] [Q&A]
## Description

This seems similar to https://github.com/apache/couchdb/issues/2236, https://github.com/apache/couchdb/issues/1852, https://github.com/apache/couchdb/issues/1552 and https://github.com/apache/couchdb/issues/1549 but the solutions there don't seem to fix anything for me.

I have a small test database full of French conjugations (7300 records). Each conjugation should be able to be queried on the verb root, pronoun (aka `agent`), and some temporal option. 

Here is a reasonable example query:

```json
{
   "selector": {
      "input.root": "aimer",
      "input.agent": {
         "$in": [
            "2-sg",
            "1-sg"
         ]
      },
      "input.option": {
         "$in": [
            "conditionnel-present",
            "indicatif-present"
         ]
      }
   },
   "use_index": "foo-json-index"
}
```
If I run this without any indexing, the performance is (expectedly) bad and I get this helpful message:

![image](https://user-images.githubusercontent.com/14815080/80752573-8a8c9d80-8ae0-11ea-9e84-7f6b96d2c2d6.png)

So, I'm indexing on verb root like so:

```json{
 "type": "json",
 "partitioned": false,
 "def": {
  "fields": [
   {
    "input.root": "asc"
   }
  ]
 }
```

If I run the query again, this dramatically improves the search time:

![image](https://user-images.githubusercontent.com/14815080/80752680-c0ca1d00-8ae0-11ea-981d-b5c5167f90e6.png)

However, if I then use the `$in` operator (I also tried the `$or` operator) then the search time skyrockets to 4-6 seconds:

![image](https://user-images.githubusercontent.com/14815080/80752786-fa9b2380-8ae0-11ea-9f41-6c723f40eca6.png)

## Steps to Reproduce

I supplied the steps above - I am using Fauxton for all of my testing/debugging.

Here is the `explain` json for the query:

```json
{
 "dbname": "data",
 "index": {
  "ddoc": "_design/3d80683e2f768bc271c5998a4bd8f03f7fe3802c",
  "name": "foo-json-index",
  "type": "json",
  "partitioned": false,
  "def": {
   "fields": [
    {
     "input.root": "asc"
    }
   ]
  }
 },
 "partitioned": false,
 "selector": {
  "$and": [
   {
    "input.root": {
     "$in": [
      "aimer",
      "etre"
     ]
    }
   },
   {
    "input.agent": {
     "$in": [
      "2-sg",
      "1-sg"
     ]
    }
   },
   {
    "input.option": {
     "$in": [
      "conditionnel-present",
      "indicatif-present"
     ]
    }
   }
  ]
 },
 "opts": {
  "use_index": [],
  "bookmark": "nil",
  "limit": 25,
  "skip": 0,
  "sort": {},
  "fields": "all_fields",
  "partition": "",
  "r": [
   49
  ],
  "conflicts": false,
  "stale": false,
  "update": true,
  "stable": false,
  "execution_stats": false
 },
 "limit": 25,
 "skip": 0,
 "fields": "all_fields",
 "mrargs": {
  "include_docs": true,
  "view_type": "map",
  "reduce": false,
  "partition": null,
  "start_key": [],
  "end_key": [
   "<MAX>"
  ],
  "direction": "fwd",
  "stable": false,
  "update": true,
  "conflicts": "undefined"
 }
}
```

Note that the `start_key` is empty and the `end_key` is `<MAX>`. If it's just one root being looked up, I instead get 
```json
"start_key": [
   "aimer"
  ],
  "end_key": [
   "aimer",
   "<MAX>"
  ]
```

## Expected Behaviour

I would expect that there would be some interpretation of the query to first pull each indexed `root` and then run the remaining `option` and `agent` conditions on those indices, but instead, it appears to just run all of these conditions on every record in the database. 

To hack my way around this (following https://github.com/apache/couchdb/issues/2236#issuecomment-546883183), if my api receives multiple `root` values, I split them up, run individual queries and then join them before returning the response, but surely there must be a more relaxing way :)

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )
[TIP]:  # ( You can paste the output of curl http://YOUR-COUCHDB:5984/ here. )

```json
couchdb: "Welcome",
version: "3.0.0",
git_sha: "03a77db6c",
uuid: "35461b15ec6019d5ff392ef60b9d8a82",
features: [
"access-ready",
"partitioned",
"pluggable-storage-engines",
"reshard",
"scheduler"
],
vendor: {
name: "The Apache Software Foundation"
}
}
```

* Browser name and version: Chrome Version 81.0.4044.129
* Operating system and version: Mac OS 10.14.6

**Comment by wohali** [comment]
Can you share your specific index definition?

**Comment by wohali** [comment]
This looks like a duplicate of #2236 on further reflection. @willholley 's response there is accurate:

> The query engine will iterate through the index with include_docs=true and evaluate each document individually against the query selector in series.

And as @garrensmith says:

> An $or operator also can't always use an index unless both sides of the $or have the same fields.

I don't think we can improve this, but I'll leave it open for a bit to see if anyone has any ideas.

**Comment by roedoejet** [comment]
> Can you share your specific index definition?

```json
{
   "index": {
      "fields": [
         "input.root"
      ]
   },
   "name": "foo-json-index",
   "type": "json"
}
```

**Comment by roedoejet** [comment]
> An $or operator also can't always use an index unless both sides of the $or have the same fields.

Unless I'm misunderstanding something, I thought that's what I tried with `$or`:

```json
{
   "selector": {
      "$or": [
         {
            "input.root": "aimer"
         },
         {
            "input.root": "etre"
         }
      ],
      "input.agent": {
         "$in": [
            "2-sg",
            "1-sg"
         ]
      },
      "input.option": {
         "$in": [
            "conditionnel-present",
            "indicatif-present"
         ]
      }
   }
}
```

But the same thing happens, it appears to be searching through each record:

![image](https://user-images.githubusercontent.com/14815080/80758249-40101e80-8aea-11ea-90a7-21ebc6a5b740.png)

**Comment by wohali** [comment]
Every field in your query must be contained in the index.

**Comment by roedoejet** [comment]
> Every field in your query must be contained in the index.

In order to use `$or` or `$in` every field in the query must be indexed? 

If I change my index to:
```json
{
   "index": {
      "fields": [
         "input.root",
         "input.agent",
         "input.option"
      ]
   },
   "name": "root-agent-option-json-index",
   "type": "json"
}
```

It still doesn't use my index:

![image](https://user-images.githubusercontent.com/14815080/80758881-36d38180-8aeb-11ea-9921-40d082eb59c0.png)

**Comment by wohali** [comment]
I can't speak to the or/in portion specifically (I believe the comments from the other devs indicate you'll never get this work for those), but I can say that if your selector included `input.root`, `input.agent`, and `input.option` (as static single values) your index must also include all 3 of those fields for it to be selected at query time. If I understand correctly, that is :)

**Comment by roedoejet** [comment]
> your index must also include all 3 of those fields for it to be selected at query time

Hm, OK, in that case I guess I'm confused as to why adding the index for just `input.root` made such a drastic improvement on my initial query:

#### Index

```json
{
   "index": {
      "fields": [
         "root.input"
      ]
   },
   "name": "foo-json-index",
   "type": "json"
}
```

#### Before adding Index above

![image](https://user-images.githubusercontent.com/14815080/80767279-68087d80-8afc-11ea-8cc5-884a2f883766.png)

#### After adding Index above

![image](https://user-images.githubusercontent.com/14815080/80767327-7d7da780-8afc-11ea-8595-2f2587aeb719.png)

**Comment by wohali** [comment]
I'll hand it over to @garrensmith or @willholley who are in UTC-ish timezones and will respond tomorrow.

**Comment by roedoejet** [comment]
> I'll hand it over to @garrensmith or @willholley who are in UTC-ish timezones and will respond tomorrow.

Thanks, I really appreciate your time looking at this!

**Comment by willholley** [comment]
@roedoejet `json` indexes basically don't have any support for the `$in` operator today - they require some kind of range query in the selector in order to use an index.

One option is to install the [text index plugin](https://docs.couchdb.org/en/master/ddocs/search.html), which will give you native `$in` support via Lucene indexes.

**Comment by roedoejet** [comment]
> @roedoejet `json` indexes basically don't have any support for the `$in` operator today - they require some kind of range query in the selector in order to use an index.
> 
> One option is to install the [text index plugin](https://docs.couchdb.org/en/master/ddocs/search.html), which will give you native `$in` support via Lucene indexes.

OK, thanks for this. I would love to see this documented [here](https://docs.couchdb.org/en/2.2.0/api/database/find.html#condition-operators). There is already a note about the regex operator not accepting indices, so I assumed that was the only operator to note wrt indices.

What about my example with `$or`?

**Comment by willholley** [comment]
The basic behaviour for `json` indexes is that if an `$or` operator is used and there is no other operator that allows an index to be selected, a full database scan is performed. There is no support for parallel or even multiple index scans when servicing a single query. 

iirc (@garrensmith can confirm, there is an optimization, whereby if all of the `$or` clauses reference the same field (or set of fields), then an index containing exactly those fields will be used in preference to a full database scan. This is because the index is guaranteed to include all the documents in the database with those fields present.

Text indexes can again be more efficient for this as they have richer support for disjoint selectors.

**Comment by nkev** [comment]
For reference, I found [this](https://github.com/apache/couchdb/issues/1853#issuecomment-454285236) by @garrensmith where he says the $in operator **does** use an index in some cases. It would be nice to get an update on this.

**Comment by garrensmith** [comment]
The `$in` can be used in an index. Say for the example of the index:

```
{
   "index": {
      "fields": [
         "root.input"
      ]
   },
   "name": "foo-json-index",
   "type": "json"
}
```
With selector:
```
{
   selector: {
     input.root: "aimer",
     input.option: {
       $in: ["option1", "option2"]
    }
   }
}
```
Mango will choose the above index because you have `input.root` in your selector. Then when it reads the docs from the index it will filter them out based on your fields in the `$in` section. 

In another case if you had an index like this:

```
{
   "index": {
      "fields": [
         "input.option"
      ]
   },
   "name": "foo-json-index",
   "type": "json"
}
```

And the same query:

```
{
   selector: {
     input.root: "aimer",
     input.option: {
       $in: ["option1", "option2"]
    }
   }
}
```

Then mango will read everything from the `input.option` index and then filter each document with the full selector. 
The better option is to have the first index as it will reduce the number of documents that need to be fetched from the index.
Hope that clears it up a bit more.

**Reply by nkev** [reply]
Thanks for the clarification Garren. One thing though, you have `root.input` in the first code snippet and `input.root` afterwards. Is that a typo?
Also, will this all be the same in CouchDB V4?

**Comment by garrensmith** [comment]
Yes it will

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
